config/backend-config/internal/cache/cache.go

```diff
@@ -0,0 +1,156 @@
+//go:generate mockgen -destination=./mock_cache.go -package=cache -source=./cache.go cache
+package cache
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+
+	"github.com/rudderlabs/rudder-server/config"
+	migrator "github.com/rudderlabs/rudder-server/services/sql-migrator"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	"github.com/rudderlabs/rudder-server/utils/pubsub"
+)
+
+var pkgLogger = logger.NewLogger().Child("backend-config-cache")
+
+type Cache interface {
+	Get(ctx context.Context) ([]byte, error)
+}
+
+type cacheStore struct {
+	*sql.DB
+	secret string
+	key    string
+}
+
+// Start returns a new Cache instance, and starts a goroutine to cache the config
+//
+// secret is the secret key to encrypt the config with before storing it
+//
+// key is the key to use to store and fetch the config from the cache store
+//
+// ch is the channel to listen on for config updates and store them
+func Start(ctx context.Context, secret, key string, ch pubsub.DataChannel) Cache {
+	var (
+		err    error
+		dbConn *sql.DB
+	)
+	// setup db connection
+	dbConn, err = setupDBConn()
+	dbStore := cacheStore{
+		dbConn,
+		secret,
+		key,
+	}
+	if err != nil {
+		pkgLogger.Errorf("failed to setup db: %v", err)
+		return &dbStore
+	}
+
+	// apply migrations
+	err = migrate(dbConn)
+	if err != nil {
+		pkgLogger.Errorf("failed to migrate db: %v", err)
+		return &dbStore
+	}
+
+	// clear config for other keys
+	err = dbStore.clear(ctx)
+	if err != nil {
+		pkgLogger.Errorf("failed to clear previous config: %v", err)
+		return &dbStore
+	}
+
+	go func() {
+		// subscribe to config and write to db
+		for config := range ch {
+			// persist to database
+			err = dbStore.set(ctx, config.Data)
+			if err != nil {
+				pkgLogger.Errorf("failed writing config to database: %v", err)
+			}
+		}
+		dbStore.Close()
+	}()
+	return &dbStore
+}
+
+// Encrypt and store the config to the database
+func (db *cacheStore) set(ctx context.Context, config interface{}) error {
+	configBytes, err := json.Marshal(config)
+	if err != nil {
+		return err
+	}
+	// write to config table
+	_, err = db.ExecContext(
+		ctx,
+		`INSERT INTO config_cache (key, config) VALUES ($1, pgp_sym_encrypt($2, $3))
+		on conflict (key)
+		do update set
+		config = pgp_sym_encrypt($2, $3),
+		updated_at = NOW()`,
+		db.key,
+		configBytes,
```

Comment:
It might be slightly better if the encryption was happening client side:
1. The connection to DB is not happening on a secure channel (SSL is not used at the moment); an network ease dropper could get access to configs
2. It would put less pressure on DB doing the compute here, although sym encryption is relatively light.
3. If a bug appears in the encryption function is easier to upgrade the code than the DB.

Reply:
one suggestion was to use [AES Encryption/Decryption in GoLang](https://golangdocs.com/aes-encryption-decryption-in-golang) to encrypt the payload in the server and then persist to database.
Any other approaches you have in mind @lvrach ?
cc @atzoum